Lib/pathlib.py

```diff
@@ -1173,6 +1212,209 @@ def walk(self, top_down=True, on_error=None, follow_symlinks=False):
 
             paths += [path._make_child_relpath(d) for d in reversed(dirnames)]
 
+    def absolute(self):
+        """Return an absolute version of this path
+        No normalization or symlink resolution is performed.
+
+        Use resolve() to resolve symlinks and remove '..' segments.
+        """
+        raise UnsupportedOperation(f"{type(self).__name__}.absolute()")
+
+    @classmethod
+    def cwd(cls):
+        """Return a new path pointing to the current working directory."""
+        return cls().absolute()
+
+    def expanduser(self):
+        """ Return a new path with expanded ~ and ~user constructs
+        (as returned by os.path.expanduser)
+        """
+        raise UnsupportedOperation(f"{type(self).__name__}.expanduser()")
+
+    @classmethod
+    def home(cls):
+        """Return a new path pointing to expanduser('~').
+        """
+        return cls("~").expanduser()
+
+    def readlink(self):
+        """
+        Return the path to which the symbolic link points.
+        """
+        raise UnsupportedOperation(f"{type(self).__name__}.readlink()")
+
+    def resolve(self, strict=False):
+        """
+        Make the path absolute, resolving all symlinks on the way and also
+        normalizing it.
+        """
+        try:
+            path = self.absolute()
+            tail_idx = len(path._tail) - len(self._tail)
+        except UnsupportedOperation:
+            path = self
+            tail_idx = 0
+        if not path._tail:
+            return path
+        drv = path.drive
+        root = path.root
+        tail = list(path._tail)
+        dirty = False
+        link_count = 0
+        readlink_supported = True
+        while tail_idx < len(tail):
+            if tail[tail_idx] == '..':
+                if tail_idx == 0:
+                    if root:
+                        # Delete '..' part immediately following root.
+                        del tail[tail_idx]
+                        dirty = True
+                        continue
+                elif tail[tail_idx - 1] != '..':
+                    # Delete '..' part and its predecessor.
+                    tail_idx -= 1
+                    del tail[tail_idx:tail_idx + 2]
+                    dirty = True
+                    continue
+            elif readlink_supported:
+                link = self._from_parsed_parts(drv, root, tail[:tail_idx + 1])
+                try:
+                    link_target = link.readlink()
+                except UnsupportedOperation:
+                    readlink_supported = False
+                except OSError as e:
+                    if e.errno != EINVAL:
+                        if strict:
+                            raise
+                        else:
+                            break
+                else:
+                    link_count += 1
+                    if link_count >= 40:
```

Comment:
why this hard-coded limit ? 

I mean why not keep a list/set of already resolved link path here and check if we see one already existing ?

Reply:
> why this hard-coded limit ?

Simply because that's what Linux does! 

https://github.com/torvalds/linux/blob/4fb0dacb78c6a041bbd38ddd998df806af5c2c69/include/linux/namei.h#L13

https://github.com/torvalds/linux/blob/master/fs/namei.c#L1728-L1729

> I mean why not keep a list/set of already resolved link path here and check if we see one already existing ?

It looks like it's not particularly easy to detect symlink loops: https://unix.stackexchange.com/questions/99159/is-there-an-algorithm-to-decide-if-a-symlink-loops

Reply:
> Simply because that's what Linux does!

Python is not Linux only ! ;)

```python
# before the while loop:
links_seen = set()

# then here:
s = str(link)
if s in links_seen:
    raise OSError(...)
links_seen.add(s)
```

?

Reply:
It's a bit more subtle than simply bailing out if we see the same stat twice. This is a symlink loop by some definitions, but `resolve()` should be able to handle it as `realpath` does:

```sh
barney@acorn ~/projects/cpython/testdir/c $ ln -s . a
barney@acorn ~/projects/cpython/testdir/c $ ln -s . b
barney@acorn ~/projects/cpython/testdir/c $ realpath a/b/a/b/a/b
/home/barney/projects/cpython/testdir/c
```

Similarly, `glob.glob()` and `pathlib.Path.glob()` will expand these paths up to 40 levels:

```python
>>> g = pathlib.Path().glob('**/', follow_symlinks=True)
>>> next(g)
PosixPath('.')
>>> next(g)
PosixPath('b')
>>> next(g)
PosixPath('b/b')
>>> next(g)
PosixPath('b/b/b')
>>> next(g)
PosixPath('b/b/b/b')
```

(I suppose this generator will have around 2^40 items)

Also, this is rather performance-sensitive code, and I want to optimize for the common case where we _don't_ need to raise `ELOOP`. Counting symlinks, as Linux does, is about the cheapest implementation possible. It's performance sensitive because any user subclass of `PathBase` that supports symlinks will need to call `resolve()` in its most fundamental methods, like `stat()`, `iterdir()` and `open()`. For example, the implementation of `DummyPath.iterdir()` in the pathlib tests reads:

```python
    def iterdir(self):
        path = str(self.resolve())
        if path in self._files:
            raise NotADirectoryError(errno.ENOTDIR, "Not a directory", path)
        elif path in self._directories:
            return (self / name for name in self._directories[path])
        else:
            raise FileNotFoundError(errno.ENOENT, "File not found", path)
```

Reply:
I'm struggling to dig out macOS docs, but [this page](https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man2/open.2.html) says:

```
     [ELOOP]            Too many symbolic links are encountered in translating
                        the pathname.  This is taken to be indicative of a
                        looping symbolic link.
```

So I think Mac uses a similar algorithm.

Reply:
I've adjusted the exception message to read "Too many symbolic links in path", and stored the limit as `pathlib._MAX_SYMLINKS`.